crates/bevy_ecs/src/schedule/stepping.rs

```diff
@@ -0,0 +1,1409 @@
+                    match self.schedule_states.get_mut(&label) {
+                        Some(state) => state.clear_behavior(system),
+                        None => {
+                            warn!(
+                                "stepping is not enabled for schedule {:?}; \
+                                use `.add_stepping({:?})` to enable stepping",
+                                label, label
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        if reset_cursor {
+            self.reset_cursor();
+        }
+    }
+
+    /// get the list of systems this schedule should skip for this render
+    /// frame
+    pub fn skipped_systems(&mut self, schedule: &Schedule) -> Option<FixedBitSet> {
+        if self.action == Action::RunAll {
+            return None;
+        }
+
+        // grab the label and state for this schedule
+        let label = schedule.name();
+        let state = self.schedule_states.get_mut(&label)?;
+
+        // Stepping is enabled, and this schedule is supposed to be stepped.
+        //
+        // We need to maintain a list of schedules in the order that they call
+        // this function. We'll check the ordered list now to see if this
+        // schedule is present. If not, we'll add it after the last schedule
+        // that called this function. Finally we want to save off the index of
+        // this schedule in the ordered schedule list. This is used to
+        // determine if this is the schedule the cursor is pointed at.
+        let index = self.schedule_order.iter().position(|l| *l == label);
+        let index = match (index, self.previous_schedule) {
+            (Some(index), _) => index,
+            (None, None) => {
+                self.schedule_order.insert(0, label);
+                0
+            }
+            (None, Some(last)) => {
+                self.schedule_order.insert(last + 1, label);
+                last + 1
+            }
+        };
+        // Update the index of the previous schedule to be the index of this
+        // schedule for the next call
+        self.previous_schedule = Some(index);
+
+        #[cfg(test)]
+        debug!(
+            "cursor {:?}, index {}, label {:?}",
+            self.cursor, index, label
+        );
+
+        // if the stepping frame cursor is pointing at this schedule, we'll run
+        // the schedule with the current stepping action.  If this is not the
+        // cursor schedule, we'll run the schedule with the waiting action.
+        let cursor = self.cursor;
+        let (list, next_system) = if index == cursor.schedule {
+            let o = state.skipped_systems(schedule, cursor.system, self.action);
+
+            // if we just stepped this schedule, then we'll switch the action
+            // to be waiting
+            if self.action == Action::Step {
+                self.action = Action::Waiting;
+            }
+            o
```

Comment:
either do this or expand `o` to something like `(skip_list, next_system)`
```suggestion
            let action = self.action;

            // if we just stepped this schedule, then we'll switch the action
            // to be waiting
            if action == Action::Step {
                self.action = Action::Waiting;
            }
            state.skipped_systems(schedule, cursor.system, action)
```

Reply:
Agreed that `o` is not the appropriate variable name here.

I'm uncomfortable flipping the action change to before the call to `skipped_systems()` because it's a break from the logical order of what is happening.  I even had a moment where I missed the save to `action` and thought the change wouldn't work.

I'll split the result out into `(skip_list, next_system)` to match the `else` block, and standardize across that whole function on those names.